meld/cluster_topdown.py

```diff
@@ -0,0 +1,305 @@
+                                      self.h_plus1,
+                                      lower_limit)
+            self.accum_val += new_slice_val
+            self.upper_limit = lower_limit
+
+    def add_member(self, node, height, extent):
+        """
+            Add a new node to cluster
+
+            Args
+            ----
+            node : int
+                Node index
+            height : float
+                Test statistic value at node
+                called "val" in Cpp implementation
+            extent : float
+                Extent of node
+                called "area" in Cpp implementation
+        """
+        self.update(height)
+        # append node to list of members
+        self.members.append(node)
+        self.extent += extent
+
+
+def alloc_cluster(cluster_list, dead_clusters, param_e, param_h):
+    """
+        Generate a valid (new) cluster number
+        If available, number for a dead cluster will be recycled
+        Cluster and dead cluster lists will be updated
+
+        Args
+        ----
+            cluster_list : list of Clusters
+            dead_clusters : set of Clusters
+            param_e : float
+            param_h : float
+    """
+    # verify that the lists are actually being modified in main code
+    if len(dead_clusters) == 0:
+        cluster_list.append(Cluster(param_e, param_h))
+        return len(cluster_list) - 1
+    else:
+        # removes and returns element from set
+        ret = dead_clusters.pop()
+        cluster_list[ret] = Cluster(param_e, param_h)
+        return ret
+
+
+def get_node_neighbors(node, connectivity):
+    """
+        Given a connectivity matrix, get neighbors of a node
+
+        Args
+        ----
+            node : int
+                Node index
+            connectivity : sparse matrix in COO format
+
+        Returns
+        -------
+            neighbors : numpy array of integers
+                array of node indices for neighboring nodes
+
+    """
+    # since connectivity matrix is triangular, we need to check both along rows
+    # and cols
+    # diagonals are all 0 in connectivity matrix (no self-neighboring) so we
+    # don't need to worry about that
+    neighbors = connectivity[node]
+
+    return neighbors
+
+
+@jit(nopython=True)
+def get_touching(neighbors, membership):
+    """ Numba function to get cluster membership of neighboring clusters
+
+    Paramters
+    ---------
+    neighbors : array of int
+        Node indicies of neighboring nodes
+    membership : array of float
+        Array storing membership for every node
+
+    Returns
+    -------
+    touching_clusters: set of int
+        Set of node indicies for touching clusters
+    """
+    touching_clusters = set()
+    for nbr in neighbors:
+        if membership[nbr] != -1:
+            touching_clusters.add(membership[nbr])
+    return touching_clusters
+
+
+def tfce_pos(col_data, area_data, connectivity, param_e=0.5, param_h=2.0):
```

Comment:
This function has a tight loop, containing conditionals, memory allocation, and function calls, so if it is possible to tell python to precompile it, that may help considerably (though it does call into classes, and I don't know the restrictions on precompilation are).